### ISSUE TITLE
Infinite loop caused by unstable context value

### ISSUE THREAD

**Issue by mischnic** [preactjs/preact#2715]
### Reproduction

https://codesandbox.io/s/preact-batch-state-context-z71k8

```js
// **** works
import { render } from "react-dom";
import * as React from "react";

// **** doesn't work:
// import { render } from "preact/compat";
// import * as React from "preact/compat";

// ----------------------------------------------------------

const Context = React.createContext(null);

function ModalProvider(props) {
	let [modalCount, setModalCount] = React.useState(0);
	let context = {
		modalCount,
		addModal() {
			setModalCount((count) => count + 1);
		},
		removeModal() {
			setModalCount((count) => count - 1);
		}
	};

	return <Context.Provider value={context}>{props.children}</Context.Provider>;
}

export function useModal() {
	let context = React.useContext(Context);
	React.useEffect(() => {
		console.log("useModal mount");

		context.addModal();
		return () => {
			console.log("useModal unmount");
			context.removeModal();
		};
	}, [context]);
}

// ----------------------------------------------------------

function Popover() {
	useModal();
	return <div>Popover</div>;
}

function App() {
	return (
		<ModalProvider>
			<Popover />
		</ModalProvider>
	);
}

render(<App />, document.getElementById("root"));
```

### Steps to reproduce

- Open the codesandbox
- Currently, it's using React which works as expected
- Change the imports to use Preact instead
- An infinite loop caused by `useEffect` and `setState` via the context occurs

### Expected Behavior

No infinite loop (as React)

### Actual Behavior

I think Preact batches updates differently, so the useEffect and useEffect cleanup callback don't cancel out.

**Comment by marvinhagemeister** [comment]
Just had a quick look at this and wrapping the context value with a `useMemo` resolves the infinite loop. This leads me to assume that React either has some infinite loop detection and bails out somewhere, or that they have special handling for objects passed into a context provider.

**Comment by LXSMNSYC** [comment]
@marvinhagemeister, if I remember correctly, React's render phase tries to detect a re-render request and throws an error after a defined number of cycles has been reached before it can go to the commit phase

**Comment by joeythomaschaske** [comment]
@marvinhagemeister Can you provide code to explain what you mean?

**Comment by mischnic** [comment]
(This was the fix that solved it for me: https://github.com/adobe/react-spectrum/commit/78ac497400786ba2572d3ee076a158038e4a409e)

```diff
 export function ModalProvider(props: ModalProviderProps) {
   let {children} = props;
   let parent = useContext(Context);
   let [modalCount, setModalCount] = useState(parent ? parent.modalCount : 0);
-  let context = {
+  let context = useMemo(() => ({
     parent,
     modalCount,
     modalCount,
     addModal() {
          setModalCount((count) => count + 1);
     },
     removeModal() {
          setModalCount((count) => count - 1);
     }
     }
-  };
+  }), [parent, modalCount]);
 
   return (
     <Context.Provider value={context}>
```

**Comment by kitten** [comment]
@marvinhagemeister It's not that special, and I believe this is related to state update batching being the default in React now. It doesn't really depend on batching even, since both the teardown and next `useEffect` run run in the same tick. The cause for React bailing out is that `setModalCount` does a `count + 1` and a `count - 1` state update, bringing it back to the original value. It hence arrives at the bail out case after seeing that the next state update is obsolete.
IMHO not a bail out case that's worth it in Preact. Two phase hooks are obviously nice and batching state updates and such and comparing state updates is useful, but it's just a micro-optimisation in most cases 😄

**Comment by mjgerace** [comment]
I don't know if this is preact's fault, but I detailed the issue with Chakra here: https://github.com/chakra-ui/chakra-ui/issues/4397#issuecomment-1041125004

**Comment by TheThirdRace** [comment]
> IMHO not a bail out case that's worth it in Preact. Two phase hooks are obviously nice and batching state updates and such and comparing state updates is useful, but it's just a micro-optimisation in most cases 😄

While I can understand the sentiment, this is the kind of reasons why people simply can't move to `Preact`.

`Preact` is awesome, I would love to be able to shed 35 KB of JS by using it, but if every now and then a UI library is using that kind of micro optimization available by default in `React` and this cause infinite loops, what am I supposed to do?

Reporting these issues to the UI library is a bit pointless, it works in `React` so it gets shoved at the bottom of the priority list. They definitely should fix it, but that's not what's happening.

If I have to choose between a much larger JS bundle by using `React` or an infinite loop that crashes my app by using `Preact`, as sad as it is, `Preact` has no chance whatsoever to win that choice...

So I would kindly ask you to reconsider your position. Even though you are 100% right, the result is not beneficial to the community and is affecting `Preact` usability in a lot of projects.

**Comment by JoviDeCroock** [comment]
We have implemented render repetition for the same component now from 10.8 and onwards, however this is still a bit of a different case as it kind off relates to two different renders.

we press addModal --> context udates --> effect triggers --> addModal gets triggered --> context updates --> ...

Basically it's fixable with the fix implemented in #3348 where we keep a `_pendingValue` for state and compare only when everything settles.

**Comment by JoviDeCroock** [comment]
Circling back to this, I do wonder what the heuristic is in React to signal to ContextConsumers as this feels like a bug to me, consider the [following sandbox](https://codesandbox.io/s/recursing-kare-e3ts3g?file=/src/index.js)

![log output showing a render with the first context value, following an add for the modal, another context value, a remove and add for the modal and then another new context value but not following up with a new remove/add](https://user-images.githubusercontent.com/17125876/175764583-753d1147-5396-433f-af28-2e24badccdfb.png)

When we consider the above log output we see that the `ModalProvider` renders three times, due to how the context value is created we have _three different values_, however we don't see three signals to the `useContext`. Afaik this was how it used to work 😅 as long as it's different it should trigger a render.
